File: tests/oracle/ptrarith_addtree_1204.cc
/*
 * Locked Ghidra 12.0.4 RulePtrArith::applyOp oracle (AddTree family).
 *
 * Every case records the target-relevant structural IR immediately before and
 * after the real rule call: block and op-bank alive/dead order, op lifecycle
 * state (opcode/dead/parent/order/inputs/output), the def-use state of every
 * fixture-tracked and transformation-created Varnode, plus the data-type
 * metatype read facing the rule. Unique-space offsets are represented by
 * stable fixture identities because they are allocation-only temporaries.
 *
 * The five cases pin the RULE-PTRARITH-ADDTREE-0001 activation chain that
 * next_url exercises end to end:
 *   - ptradd8_load          INT_ADD(char**, 0x38) feeding a LOAD pointer slot
 *                           converts to PTRADD(ptr, 7, 8) — the `ppcVar6+7`
 *                           shape of the golden output.
 *   - ptrsub_struct_load    INT_ADD(struct*, field-off) feeding a LOAD
 *                           converts to PTRSUB(ptr, off).
 *   - charpp_nonmult_no_chg a non-multiple constant on a scalar base keeps
 *                           the INT_ADD (valid=false path).
 *   - untyped_base_no_chg   an untyped base never reaches the slot search
 *                           pass (the pre-factory production failure mode).
 *   - recovery_not_started  the hasTypeRecoveryStarted() first guard.
 */

#include "bfd_arch.hh"
#include "libdecomp.hh"
#include "ruleaction.hh"

#include <iomanip>
#include <iostream>
#include <iterator>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::ostringstream;
using std::string;
using std::vector;

class Fixture {
  Funcdata &fd;
  BlockBasic *block;
  vector<PcodeOp *> ops;
  map<PcodeOp *,string> opNames;
  vector<Varnode *> varnodes;
  map<Varnode *,string> varnodeNames;
  int4 nextOp;
  int4 nextVarnode;

  void rememberOp(PcodeOp *op,const string &name)
  {
    if (opNames.insert(std::make_pair(op,name)).second)
      ops.push_back(op);
  }

  void rememberVarnode(Varnode *vn,const string &name)
  {
    if (varnodeNames.insert(std::make_pair(vn,name)).second)
      varnodes.push_back(vn);
  }

  string opName(PcodeOp *op) const
  {
    map<PcodeOp *,string>::const_iterator iter = opNames.find(op);
    if (iter == opNames.end())
      throw std::runtime_error("unregistered fixture op");
    return (*iter).second;
  }

  string varnodeName(Varnode *vn) const
  {
    map<Varnode *,string>::const_iterator iter = varnodeNames.find(vn);
    if (iter == varnodeNames.end())
      throw std::runtime_error("unregistered fixture varnode");
    return (*iter).second;
  }

  void discover(void)
  {
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();
        iter!=block->endOp();++iter) {
      if (opNames.find(*iter) == opNames.end()) {
        ostringstream name;
        name << 'n' << nextOp++;
        rememberOp(*iter,name.str());
      }
    }
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();
        iter!=block->endOp();++iter) {
      PcodeOp *op = *iter;
      if (op->getOut() != (Varnode *)0 &&
          varnodeNames.find(op->getOut()) == varnodeNames.end())
        rememberVarnode(op->getOut(),opName(op) + "_out");
      for(int4 slot=0;slot<op->numInput();++slot) {
        Varnode *vn = op->getIn(slot);
        if (varnodeNames.find(vn) == varnodeNames.end()) {
          ostringstream name;
          name << 'g' << nextVarnode++;
          rememberVarnode(vn,name.str());
        }
      }
    }
  }

  string opList(list<PcodeOp *>::const_iterator iter,
                list<PcodeOp *>::const_iterator enditer) const
  {
    ostringstream out;
    bool first = true;
    for(;iter!=enditer;++iter) {
      if (!first) out << ',';
      first = false;
      out << opName(*iter);
    }
    return out.str();
  }

  string blockOrder(void) const
  {
    return opList(block->beginOp(),block->endOp());
  }

  string opState(PcodeOp *op) const
  {
    ostringstream out;
    out << opName(op)
        << "{opc=" << static_cast<int4>(op->code())
        << ",dead=" << op->isDead()
        << ",parent=" << (op->getParent() == block ? block->getIndex() : -1)
        << ",order=" << op->getSeqNum().getOrder()
        << ",inputs=[";
    for(int4 slot=0;slot<op->numInput();++slot) {
      if (slot != 0) out << ',';
      out << varnodeName(op->getIn(slot));
    }
    out << "],output=";
    if (op->getOut() == (Varnode *)0)
      out << '-';
    else
      out << varnodeName(op->getOut());
    out << '}';
    return out.str();
  }

  string typeTag(Varnode *vn) const
  {
    const Datatype *ct = vn->getType();
    if (ct == (const Datatype *)0 || ct->getMetatype() == TYPE_VOID)
      return "void";
    ostringstream out;
    switch(ct->getMetatype()) {
    case TYPE_PTR: out << "ptr"; break;
    case TYPE_STRUCT: out << "struct"; break;
    case TYPE_ARRAY: out << "array"; break;
    case TYPE_INT: out << "int"; break;
    case TYPE_UINT: out << "uint"; break;
    case TYPE_UNKNOWN: out << "unknown"; break;
    default: out << "other"; break;
    }
    out << ':' << ct->getSize();
    return out.str();
  }

  string varnodeState(Varnode *vn) const
  {
    ostringstream out;
    out << varnodeName(vn)
        << "{space=" << vn->getSpace()->getName()
        << ",size=" << vn->getSize()
        << ",offset=";
    if (vn->getSpace()->getType() == IPTR_INTERNAL)
      out << "tmp";
    else
      out << std::hex << vn->getOffset() << std::dec;
    out << ",constant=" << vn->isConstant()
        << ",input=" << vn->isInput()
        << ",written=" << vn->isWritten()
        << ",free=" << vn->isFree()
        << ",def=";
    if (vn->getDef() == (PcodeOp *)0)
      out << '-';
    else
      out << opName(vn->getDef());
    out << ",type=" << typeTag(vn);
    out << ",desc=[";
    bool first = true;
    for(list<PcodeOp *>::const_iterator iter=vn->beginDescend();
        iter!=vn->endDescend();++iter) {
      if (!first) out << ',';
      first = false;
      out << opName(*iter) << '.' << (*iter)->getSlot(vn);
    }
    out << "]}";
    return out.str();
  }

public:
  explicit Fixture(Funcdata &func)
    : fd(func), block((BlockBasic *)0), nextOp(0), nextVarnode(0)
  {
    BlockGraph &graph = const_cast<BlockGraph &>(fd.getBasicBlocks());
    block = graph.newBlockBasic(&fd);
  }

  Varnode *makeInput(const string &name,int4 size,uintb offset)
  {
    AddrSpace *registerSpace = fd.getArch()->getSpaceByName("register");
    if (registerSpace == (AddrSpace *)0)
      throw std::runtime_error("fixture requires the register space");
    Varnode *vn = fd.setInputVarnode(fd.newVarnode(size,registerSpace,offset));
    rememberVarnode(vn,name);
    return vn;
  }

  Varnode *makeConstant(const string &name,int4 size,uintb value)
  {
    Varnode *vn = fd.newConstant(size,value);
    rememberVarnode(vn,name);
    return vn;
  }

  PcodeOp *makeOp(const string &name,OpCode opcode,int4 inputs,int4 outputSize)
  {
    AddrSpace *codeSpace = fd.getArch()->getDefaultCodeSpace();
    PcodeOp *op = fd.newOp(inputs,Address(codeSpace,0x1000));
    fd.opSetOpcode(op,opcode);
    rememberOp(op,name);
    if (outputSize > 0) {
      Varnode *outvn = fd.newUniqueOut(outputSize,op);
      rememberVarnode(outvn,name + "_out");
    }
    return op;
  }

  void setInput(PcodeOp *op,Varnode *vn,int4 slot)
  {
    fd.opSetInput(op,vn,slot);
  }

  void insertEnd(PcodeOp *op)
  {
    fd.opInsertEnd(op,block);
  }

  void startTypeRecovery(void)
  {
    fd.startTypeRecovery();
  }

  void dump(const string &caseName,const string &stage,const string &result)
  {
    discover();
    ostringstream opStates;
    bool first = true;
    for(list<PcodeOp *>::const_iterator iter=block->beginOp();
        iter!=block->endOp();++iter) {
      if (!first) opStates << ';';
      first = false;
      opStates << opState(*iter);
    }

    ostringstream varnodeStates;
    first = true;
    for(vector<Varnode *>::const_iterator iter=varnodes.begin();
        iter!=varnodes.end();++iter) {
      if (!first) varnodeStates << ';';
      first = false;
      varnodeStates << varnodeState(*iter);
    }

    std::cout << "case=" << caseName
              << "|stage=" << stage
              << "|result=" << result
              << "|block=[" << blockOrder() << ']'
              << "|alive=[" << opList(fd.beginOpAlive(),fd.endOpAlive()) << ']'
              << "|dead=[" << opList(fd.beginOpDead(),fd.endOpDead()) << ']'
              << "|ops=[" << opStates.str() << ']'
              << "|varnodes=[" << varnodeStates.str() << "]\n";
  }
};

/// Build the AddTree shape under test: a typed base pointer input, an
/// INT_ADD with a constant, and a LOAD consuming the INT_ADD output as its
/// pointer (evaluatePointerExpression's res==2 LOAD branch). The LOAD space
/// constant uses the fixed sentinel 1: its value is never consulted by
/// RulePtrArith, and encoding the AddrSpace pointer would make the dump
/// address-layout dependent.
void runAddLoadCase(Funcdata &fd,const string &name,Datatype *baseType,
                    uintb addConstant,int4 loadSize,bool startRecovery)
{
  fd.clear();
  Fixture fixture(fd);
  TypeFactory *types = fd.getArch()->types;
  Datatype *pointerType = types->getTypePointer(baseType->getSize(),baseType,1);

  Varnode *ptr = fixture.makeInput("ptr",8,0x40);
  ptr->updateType(pointerType);
  PcodeOp *add = fixture.makeOp("add",CPUI_INT_ADD,2,8);
  Varnode *constant = fixture.makeConstant("off",8,addConstant);
  PcodeOp *load = fixture.makeOp("load",CPUI_LOAD,2,loadSize);

  fixture.setInput(add,ptr,0);
  fixture.setInput(add,constant,1);
  fixture.setInput(load,fd.newConstant(8,1),0);
  fixture.setInput(load,add->getOut(),1);
  fixture.insertEnd(add);
  fixture.insertEnd(load);
  if (startRecovery)
    fixture.startTypeRecovery();

  fixture.dump(name,"before","na");
  RulePtrArith rule("typerecovery");
  int4 result = rule.applyOp(add,fd);
  fixture.dump(name,"after",std::to_string(result));
}

/// Untyped base: the applyOp pointer-slot search must reject the op before
/// any AddTree analysis (the production pre-factory failure mode).
void runUntypedCase(Funcdata &fd,const string &name)
{
  fd.clear();
  Fixture fixture(fd);
  Datatype *intType = fd.getArch()->types->getBase(8,TYPE_INT);

  Varnode *ptr = fixture.makeInput("ptr",8,0x40);
  ptr->updateType(intType);
  PcodeOp *add = fixture.makeOp("add",CPUI_INT_ADD,2,8);
  Varnode *constant = fixture.makeConstant("off",8,0x38);
  PcodeOp *load = fixture.makeOp("load",CPUI_LOAD,2,8);

  fixture.setInput(add,ptr,0);
  fixture.setInput(add,constant,1);
  fixture.setInput(load,fd.newConstant(8,1),0);
  fixture.setInput(load,add->getOut(),1);
  fixture.insertEnd(add);
  fixture.insertEnd(load);
  fixture.startTypeRecovery();

  fixture.dump(name,"before","na");
  RulePtrArith rule("typerecovery");
  int4 result = rule.applyOp(add,fd);
  fixture.dump(name,"after",std::to_string(result));
}

void run(const string &specDirectory,const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    BfdArchitecture architecture(binary,"default",&std::cerr);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0)
      throw std::runtime_error("GetStr was not found in the BFD symbol table");
    if (fd->getName() != "GetStr" ||
        fd->getAddress().getOffset() != 0x36d0 || fd->getSize() != 0)
      throw std::runtime_error("GetStr input identity drifted");
    if (architecture.archid != "x86:LE:64:default:gcc")
      throw std::runtime_error("runtime architecture/compiler drifted: " +
                               architecture.archid);

    TypeFactory *types = architecture.types;
    TypeStruct *twoField = types->getTypeStruct("PtrarithStruct");
    {
      vector<TypeField> fields;
      Datatype *uint8Type = types->getBase(8,TYPE_UINT);
      fields.push_back(TypeField(0,0,"first",uint8Type));
      fields.push_back(TypeField(1,8,"second",uint8Type));
      types->setFields(fields,twoField,16,8,0);
    }
    Datatype *uint8Base = types->getBase(8,TYPE_UINT);

    // hasTypeRecoveryStarted() first guard: identical shape to the PTRADD
    // case but recovery never starts, so the rule must return 0 untouched.
    runAddLoadCase(*fd,"recovery_not_started",uint8Base,0x38,8,false);
    // Untyped base: slot search rejection.
    runUntypedCase(*fd,"untyped_base_no_chg");
    // char** + 0x38 -> PTRADD(ptr, 7, 8): the next_url `ppcVar6+7` shape.
    runAddLoadCase(*fd,"ptradd8_load",uint8Base,0x38,8,true);
    // struct* + field offset 8 -> PTRSUB(ptr, 8).
    runAddLoadCase(*fd,"ptrsub_struct_load",twoField,8,8,true);
    // char** + 0x3a: non-multiple remainder on a scalar base -> valid=false,
    // INT_ADD must survive untouched.
    runAddLoadCase(*fd,"charpp_nonmult_no_chg",uint8Base,0x3a,8,true);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc,char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: ptrarith_addtree_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1],argv[2]);
    return 0;
  }
  catch(const ghidra::LowlevelError &error) {
    std::cerr << "Ghidra LowlevelError: " << error.explain << '\n';
  }
  catch(const ghidra::DecoderError &error) {
    std::cerr << "Ghidra DecoderError: " << error.explain << '\n';
  }
  catch(const std::exception &error) {
    std::cerr << "fixture error: " << error.what() << '\n';
  }
  return 1;
}

File: tests/oracle/ptrarith_addtree_1204.rs
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use rugra::action::Rule;
use rugra::address::Address;
use rugra::funcdata::Funcdata;
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::ruleaction::RulePtrArith;
use rugra::space::AddressSpace;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeField, TypeMetatype, TypePointer, TypeStruct};
use rugra::varnode::Varnode;

type VarnodeRef = Arc<RwLock<Varnode>>;

/// Deterministic data-types mirroring the BfdArchitecture TypeFactory side:
/// `uint8` (8-byte unsigned), `int8` (8-byte signed) and
/// `struct PtrarithStruct { uint8 first @0; uint8 second @8; }` (size 16,
/// align 8, matching types->setFields(..., 16, 8, 0)).
fn uint8_type() -> Arc<Datatype> {
    let mut base = TypeBase::new("ulong".to_string(), 8, TypeMetatype::Uint);
    base.alignment = 8;
    base.align_size = 8;
    Arc::new(Datatype::Base(base))
}

fn int8_type() -> Arc<Datatype> {
    let mut base = TypeBase::new("long".to_string(), 8, TypeMetatype::Int);
    base.alignment = 8;
    base.align_size = 8;
    Arc::new(Datatype::Base(base))
}

fn two_field_struct() -> Arc<Datatype> {
    let uint8 = uint8_type();
    let mut base = TypeBase::new("PtrarithStruct".to_string(), 16, TypeMetatype::Struct);
    base.alignment = 8;
    base.align_size = 16;
    Arc::new(Datatype::Struct(TypeStruct {
        base,
        fields: vec![
            TypeField { name: "first".to_string(), offset: 0, type_ptr: uint8.clone() },
            TypeField { name: "second".to_string(), offset: 8, type_ptr: uint8 },
        ],
    }))
}

/// Pointer to `base_type` in byte units (wordsize 1), mirroring
/// `types->getTypePointer(baseType->getSize(), baseType, 1)`.
fn pointer_to(base_type: Arc<Datatype>) -> Arc<Datatype> {
    Arc::new(Datatype::Pointer(TypePointer::new(
        base_type.get_size(),
        base_type,
        1,
    )))
}

struct Fixture {
    fd: Funcdata,
    block: Arc<RwLock<dyn rugra::block::FlowBlock + Send + Sync>>,
    ops: Vec<PcodeOpRef>,
    op_names: HashMap<usize, String>,
    varnodes: Vec<VarnodeRef>,
    varnode_names: HashMap<usize, String>,
    next_op: usize,
    next_varnode: usize,
}

impl Fixture {
    fn new() -> Self {
        let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 0);
        assert_eq!(fd.name, "GetStr");
        assert_eq!(fd.baseaddr.as_u64(), 0x36d0);
        assert_eq!(fd.size, 0);
        let block = fd.create_new_block();
        Self {
            fd,
            block,
            ops: Vec::new(),
            op_names: HashMap::new(),
            varnodes: Vec::new(),
            varnode_names: HashMap::new(),
            next_op: 0,
            next_varnode: 0,
        }
    }

    fn op_key(op: &PcodeOpRef) -> usize {
        Arc::as_ptr(&op.0) as usize
    }

    fn op_arc_key(op: &Arc<RwLock<PcodeOp>>) -> usize {
        Arc::as_ptr(op) as usize
    }

    fn varnode_key(vn: &VarnodeRef) -> usize {
        Arc::as_ptr(vn) as usize
    }

    fn remember_op(&mut self, op: PcodeOpRef, name: String) {
        let key = Self::op_key(&op);
        if self.op_names.insert(key, name).is_none() {
            self.ops.push(op);
        }
    }

    fn remember_varnode(&mut self, vn: VarnodeRef, name: String) {
        let key = Self::varnode_key(&vn);
        if self.varnode_names.insert(key, name).is_none() {
            self.varnodes.push(vn);
        }
    }

    fn op_name(&self, op: &PcodeOpRef) -> &str {
        self.op_names
            .get(&Self::op_key(op))
            .expect("registered fixture op")
    }

    fn op_arc_name(&self, op: &Arc<RwLock<PcodeOp>>) -> &str {
        self.op_names
            .get(&Self::op_arc_key(op))
            .expect("registered fixture op")
    }

    fn varnode_name(&self, vn: &VarnodeRef) -> &str {
        self.varnode_names
            .get(&Self::varnode_key(vn))
            .expect("registered fixture varnode")
    }

    fn discover(&mut self) {
        let block_ops = self.block.read().unwrap().get_ops();
        for op in &block_ops {
            if !self.op_names.contains_key(&Self::op_key(op)) {
                let name = format!("n{}", self.next_op);
                self.next_op += 1;
                self.remember_op(op.clone(), name);
            }
        }
        for op in &block_ops {
            let op_name = self.op_name(op).to_string();
            let (output, inputs) = {
                let op = op.0.read().unwrap();
                (op.output.clone(), op.inrefs.clone())
            };
            if let Some(output) = output {
                if !self.varnode_names.contains_key(&Self::varnode_key(&output)) {
                    self.remember_varnode(output, format!("{op_name}_out"));
                }
            }
            for input in inputs {
                if !self.varnode_names.contains_key(&Self::varnode_key(&input)) {
                    let name = format!("g{}", self.next_varnode);
                    self.next_varnode += 1;
                    self.remember_varnode(input, name);
                }
            }
        }
    }

    fn make_input(&mut self, name: &str, size: usize, offset: u64) -> VarnodeRef {
        let vn = self
            .fd
            .vbank
            .create_with_space(size, AddressSpace::Register, offset);
        let vn = self.fd.set_input_varnode(vn);
        self.remember_varnode(vn.clone(), name.to_string());
        vn
    }

    fn make_constant(&mut self, name: &str, size: usize, value: u64) -> VarnodeRef {
        let vn = self.fd.new_constant(size, value);
        self.remember_varnode(vn.clone(), name.to_string());
        vn
    }

    fn make_op(
        &mut self,
        name: &str,
        opcode: OpCode,
        inputs: usize,
        output_size: usize,
    ) -> PcodeOpRef {
        let op = self.fd.new_op(inputs, Address::new(0x1000));
        self.fd.op_set_opcode(&op, opcode);
        self.remember_op(op.clone(), name.to_string());
        let output = self.fd.new_unique_out(output_size, &op);
        self.remember_varnode(output, format!("{name}_out"));
        op
    }

    fn set_input(&mut self, op: &PcodeOpRef, vn: VarnodeRef, slot: usize) {
        self.fd.op_set_input(op, vn, slot);
    }

    fn insert_end(&mut self, op: &PcodeOpRef) {
        self.fd.op_insert_end(op, &self.block.clone());
    }

    fn start_type_recovery(&mut self) {
        self.fd.set_type_recovery_started();
    }

    fn op_list(&self, ops: &[PcodeOpRef]) -> String {
        ops.iter()
            .map(|op| self.op_name(op).to_string())
            .collect::<Vec<_>>()
            .join(",")
    }

    fn op_state(&self, op: &PcodeOpRef) -> String {
        let op_guard = op.0.read().unwrap();
        let parent = op_guard
            .parent
            .as_ref()
            .and_then(std::sync::Weak::upgrade)
            .map(|parent| parent.read().unwrap().get_index())
            .unwrap_or(-1);
        let inputs = op_guard
            .inrefs
            .iter()
            .map(|vn| self.varnode_name(vn).to_string())
            .collect::<Vec<_>>()
            .join(",");
        let output = op_guard
            .output
            .as_ref()
            .map(|vn| self.varnode_name(vn).to_string())
            .unwrap_or_else(|| "-".to_string());
        format!(
            "{}{{opc={},dead={},parent={},order={},inputs=[{}],output={}}}",
            self.op_name(op),
            op_guard.opcode as i32,
            u8::from(op_guard.is_dead()),
            parent,
            op_guard.start.get_order(),
            inputs,
            output,
        )
    }

    fn type_tag(&self, vn: &VarnodeRef) -> String {
        let vn_guard = vn.read().unwrap();
        match vn_guard.get_type() {
            // Ghidra's Varnode constructor assigns a sized TYPE_UNKNOWN
            // default (glb->types->getBase(sz, TYPE_UNKNOWN)); Rugra leaves
            // v_type absent, which is the same observation.
            None => format!("unknown:{}", vn_guard.get_size()),
            Some(ct) => {
                let metatype = match ct.get_metatype() {
                    TypeMetatype::Pointer => "ptr",
                    TypeMetatype::Struct => "struct",
                    TypeMetatype::Array => "array",
                    TypeMetatype::Int => "int",
                    TypeMetatype::Uint => "uint",
                    TypeMetatype::Unknown => "unknown",
                    _ => "other",
                };
                format!("{metatype}:{}", ct.get_size())
            }
        }
    }

    fn varnode_state(&self, vn: &VarnodeRef) -> String {
        let vn_guard = vn.read().unwrap();
        let space = vn_guard.get_space();
        let offset = if space == AddressSpace::Unique {
            "tmp".to_string()
        } else {
            format!("{:x}", vn_guard.get_offset())
        };
        let definition = vn_guard
            .def
            .as_ref()
            .and_then(std::sync::Weak::upgrade)
            .map(|op| self.op_arc_name(&op).to_string())
            .unwrap_or_else(|| "-".to_string());
        let descendants = vn_guard
            .descend
            .iter()
            .filter_map(std::sync::Weak::upgrade)
            .map(|op| {
                let slot = op
                    .read()
                    .unwrap()
                    .inrefs
                    .iter()
                    .position(|input| Arc::ptr_eq(input, vn))
                    .expect("descendant retains input varnode");
                format!("{}.{}", self.op_arc_name(&op), slot)
            })
            .collect::<Vec<_>>()
            .join(",");
        let type_tag = self.type_tag(vn);
        format!(
            "{}{{space={},size={},offset={},constant={},input={},written={},free={},def={},type={},desc=[{}]}}",
            self.varnode_name(vn),
            space.name(),
            vn_guard.get_size(),
            offset,
            u8::from(vn_guard.is_constant()),
            u8::from(vn_guard.is_input()),
            u8::from(vn_guard.is_written()),
            u8::from(vn_guard.is_free()),
            definition,
            type_tag,
            descendants,
        )
    }

    fn dump(&mut self, case_name: &str, stage: &str, result: &str) {
        self.discover();
        let block_ops = self.block.read().unwrap().get_ops();
        let op_states = block_ops
            .iter()
            .map(|op| self.op_state(op))
            .collect::<Vec<_>>()
            .join(";");
        let varnode_states = self
            .varnodes
            .iter()
            .map(|vn| self.varnode_state(vn))
            .collect::<Vec<_>>()
            .join(";");
        println!(
            "case={case_name}|stage={stage}|result={result}|block=[{}]|alive=[{}]|dead=[{}]|ops=[{}]|varnodes=[{}]",
            self.op_list(&block_ops),
            self.op_list(&self.fd.obank.alivelist),
            self.op_list(&self.fd.obank.deadlist),
            op_states,
            varnode_states,
        );
    }
}

/// Mirror of the C++ runAddLoadCase: typed base pointer input, INT_ADD with a
/// constant, and a LOAD consuming the INT_ADD output as its pointer. The LOAD
/// space constant uses the fixed sentinel 1 (its value is never consulted by
/// RulePtrArith).
fn run_add_load_case(name: &str, base_type: Arc<Datatype>, add_constant: u64, start_recovery: bool) {
    let mut fixture = Fixture::new();
    let pointer_type = pointer_to(base_type);

    let ptr = fixture.make_input("ptr", 8, 0x40);
    ptr.write().unwrap().update_type(pointer_type);
    let add = fixture.make_op("add", OpCode::CPUI_INT_ADD, 2, 8);
    let constant = fixture.make_constant("off", 8, add_constant);
    let load = fixture.make_op("load", OpCode::CPUI_LOAD, 2, 8);

    fixture.set_input(&add, ptr, 0);
    fixture.set_input(&add, constant, 1);
    let space_const = fixture.fd.vbank.create_constant(8, 1);
    fixture.set_input(&load, space_const, 0);
    let add_out = add.0.read().unwrap().output.clone().unwrap();
    fixture.set_input(&load, add_out, 1);
    fixture.insert_end(&add);
    fixture.insert_end(&load);
    if start_recovery {
        fixture.start_type_recovery();
    }

    fixture.dump(name, "before", "na");
    let result = RulePtrArith::new()
        .apply_op(&add.0, &mut fixture.fd)
        .expect("RulePtrArith AddTree case");
    fixture.dump(name, "after", &result.to_string());
}

/// Mirror of the C++ runUntypedCase: an int-typed base never passes the
/// applyOp pointer-slot search.
fn run_untyped_case(name: &str) {
    let mut fixture = Fixture::new();
    let ptr = fixture.make_input("ptr", 8, 0x40);
    ptr.write().unwrap().update_type(int8_type());
    let add = fixture.make_op("add", OpCode::CPUI_INT_ADD, 2, 8);
    let constant = fixture.make_constant("off", 8, 0x38);
    let load = fixture.make_op("load", OpCode::CPUI_LOAD, 2, 8);

    fixture.set_input(&add, ptr, 0);
    fixture.set_input(&add, constant, 1);
    let space_const = fixture.fd.vbank.create_constant(8, 1);
    fixture.set_input(&load, space_const, 0);
    let add_out = add.0.read().unwrap().output.clone().unwrap();
    fixture.set_input(&load, add_out, 1);
    fixture.insert_end(&add);
    fixture.insert_end(&load);
    fixture.start_type_recovery();

    fixture.dump(name, "before", "na");
    let result = RulePtrArith::new()
        .apply_op(&add.0, &mut fixture.fd)
        .expect("RulePtrArith untyped case");
    fixture.dump(name, "after", &result.to_string());
}

fn main() {
    run_add_load_case("recovery_not_started", uint8_type(), 0x38, false);
    run_untyped_case("untyped_base_no_chg");
    run_add_load_case("ptradd8_load", uint8_type(), 0x38, true);
    run_add_load_case("ptrsub_struct_load", two_field_struct(), 8, true);
    run_add_load_case("charpp_nonmult_no_chg", uint8_type(), 0x3a, true);
}
